File: webgl/source/js/renders/modelrender.js
/**
 * For testing different models.
 */
class ModelRender extends Render {


    matrices = new Matrices();
    prog
    globject = new GlObject();
    matrix = glMatrix.mat4.create();
    rotY = 1.5 * Math.PI / 180.0;
    rotX = Math.PI / 180.0;

    lightAmbient
    lightDiffuse
    lightDirection

    onSetup(appOGL) {

        this.lightAmbient = glMatrix.vec3.fromValues(0.3, 0.3, 0.3);
        this.lightDiffuse = glMatrix.vec3.fromValues(1.0, 1.0, 1.0);
        this.lightDirection = glMatrix.vec3.fromValues(0.85, 0.8, 0.75);
        glMatrix.vec3.normalize(this.lightDirection, this.lightDirection);

        const gl = appOGL.gl;

        this.prog = new ShaderProgramBuilder(gl)
            .matrixP_V_M()
            .matrix(false)
            .colors(false, false)
            .lightDirectional(false)
            .texture2D()
            .build();

        gl.useProgram(this.prog.idProgram);

        glMatrix.mat4.translate(this.matrices.model, this.matrices.model, [0.0, 0.0, -6.0]);
        this.globject.model = Models.cube(gl, 1, 0, 0, "cube").toGlModel(gl);

        //  this.globject.model = Models.quad(gl, 2, 2, "quad").toGlModel(gl);

        this.globject.texture = GlTexture.newTexture2D(gl,
            "images/textures/235.jpg", "test");

        gl.clearColor(0.5, 0.5, 0.5, 1.0);
        gl.clearDepth(1.0); // Clear everything
        gl.enable(gl.DEPTH_TEST); // Enable depth testing
        gl.depthFunc(gl.LEQUAL); // Near things obscure far things

        console.log("max uniforms blocks", gl.getParameter(gl.MAX_COMBINED_UNIFORM_BLOCKS));
        console.log("active uniforms",
            gl.getProgramParameter(this.prog.idProgram, gl.ACTIVE_UNIFORMS));


    }


    onDrawFrame(appOGL) {
        const gl = appOGL.gl;
        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);

        glMatrix.mat4.rotate(this.matrices.model, this.matrices.model,
            this.rotX, [1, 0, 0]);
        glMatrix.mat4.rotate(this.matrices.model, this.matrices.model,
            this.rotY, [0, 1, 0]);

        gl.useProgram(this.prog.idProgram);

        this.matrices.applyCurrentModel();

        this.prog.uniformTexture(this.globject.texture);
        this.prog.uniformMatrices(this.matrices);
        this.prog.uniformDirectionalLight(this.lightAmbient,
            this.lightDiffuse, this.lightDirection);

        this.globject.model.draw(gl);
    }


}

File: lwjgl/src/main/java/com/darkraha/opengldemoj/renders/ModelRender.java
package com.darkraha.opengldemoj.renders;

import com.darkraha.opengldemoj.gl.*;
import com.darkraha.opengldemoj.gl.modelling.Models;
import org.joml.Matrix4f;
import org.joml.Vector3f;

import static org.lwjgl.opengl.GL11.*;


public class ModelRender extends Render {

    private final float rotY = 1.5f * TO_RAD;
    private final float rotX = TO_RAD;
    private final Matrices matrices = new Matrices();
    private ShaderProgram prog;
    private final GlObject object = new GlObject();

    public final Vector3f lightAmbient = new Vector3f(0.6f, 0.6f, 0.6f);
    public final Vector3f lightDiffuse = new Vector3f(1.0f, 1.0f, 1.0f);
    public final Vector3f lightDirection = new Vector3f(0.85f, 0.8f, 0.75f).normalize();


    @Override
    public void onSurfaceChanged(AppOGL appOGL, int w, int h) {
        super.onSurfaceChanged(appOGL, w, h);
        matrices.projection.identity().perspective(45 * TO_RAD, aspect, 1f, 100);

    }

    @Override
    public void onSetup(AppOGL appOGL) {

        prog = new ShaderProgramBuilder()
                .lightDirectional()
                .vertexAttributes(false, true, true)
                .matrix(true)
                .build();

        setSurfaceSize(appOGL.getWidth(), appOGL.getHeight());
        glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
        glEnable(GL_DEPTH_TEST);

        object.model = Models.sphere(1f,
                36, 36,
                1f, 0.5f, 0.5f,
                "sphere-0").toGlModel();


        object.transforms = new Matrix4f();
        object.transforms.translate(0,0,-6);
        object.texture = GlTexture.newTexture2D( "/textures/texture.png" , "terra-0");


          //      GlTexture.newSolidColorTexture("a", 0, 0, 255);

        //GlTexture.newTexture2D("terra-0", "/textures/texture.png" );
    }

    @Override
    public void onDrawFrame(AppOGL appOGL) {
//        rotY += 1.5f * TO_RAD;
//        if (rotY > 2 * PI_F) {
//            rotY = rotY - 2 * PI_F;
//        }
//
//        rotX += 1f * TO_RAD;
//        if (rotX > 2 * PI_F) {
//            rotX = rotX - 2 * PI_F;
//        }

        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
        object.transforms.rotateAffineXYZ(rotX, rotY, 0);

        matrices.applyModel(object.transforms);

        prog.use();
        prog.uniformMatrices(matrices);
        prog.uniformTexture(object.texture);
        prog.uniformDirectionalLight(lightAmbient, lightDiffuse, lightDirection);

        // glEnable(GL_CULL_FACE);
        object.model.draw();
    }
}
